RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc

```diff
@@ -0,0 +1,198 @@
+                         (calibvalid > 0));
+        bool useTOT((digi.tctp() == 3) && isAvailable);
+        bool useADC(!useTOT && isAvailable);
+        recHits[idx].energy() = useADC * adc_denoise(digi.adc(),
+                                                     digi.cm(),
+                                                     digi.adcm1(),
+                                                     calib.ADC_ped(),
+                                                     calib.CM_slope(),
+                                                     calib.CM_ped(),
+                                                     calib.BXm1_slope()) +
+                                useTOT * tot_linearization(digi.tot(),
+                                                           calib.TOT_lin(),
+                                                           calib.TOTtoADC(),
+                                                           calib.TOT_ped(),
+                                                           calib.TOT_P0(),
+                                                           calib.TOT_P1(),
+                                                           calib.TOT_P2());
+
+        //after denoising/linearization apply the MIP scale
+        recHits[idx].energy() *= calib.MIPS_scale();
+      }
+    }
+  };
+
+  //
+  struct HGCalRecHitCalibrationKernel_toaToTime {
+    template <typename TAcc>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                  HGCalDigiDevice::View digis,
+                                  HGCalRecHitDevice::View recHits,
+                                  HGCalCalibParamDevice::ConstView calibs) const {
+      auto toa_to_ps = [&](uint32_t toa, float toatops) { return toa * toatops; };
+
+      for (auto idx : uniform_elements(acc, digis.metadata().size())) {
+        auto calib = calibs[idx];
+        int calibvalid = std::to_integer<int>(calib.valid());
+        auto digi = digis[idx];
+        auto digiflags = digi.flags();
+        bool isAvailable((digiflags != hgcal::DIGI_FLAG::Invalid) && (digiflags != hgcal::DIGI_FLAG::NotAvailable) &&
+                         (calibvalid > 0));
+        bool isToAavailable((digiflags != hgcal::DIGI_FLAG::ZS_ToA) && (digiflags != hgcal::DIGI_FLAG::ZS_ToA_ADCm1));
+        bool isGood(isAvailable && isToAavailable);
+        recHits[idx].time() = isGood * toa_to_ps(digi.toa(), calib.TOAtops());
+      }
+    }
+  };
+
+  struct HGCalRecHitCalibrationKernel_printRecHits {
+    template <typename TAcc>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, HGCalRecHitDevice::ConstView view, int size) const {
+      for (int i = 0; i < size; ++i) {
+        auto const& recHit = view[i];
+        printf("%d\t%f\t%f\t%d\n", i, recHit.energy(), recHit.time(), recHit.flags());
+      }
+    }
+  };
+
+  std::unique_ptr<HGCalRecHitDevice> HGCalRecHitCalibrationAlgorithms::calibrate(
```

Comment:
Should this return by value instead of by `unique_ptr` ?
IIRC it should be guaranteed that the final copy will be elided.

Reply:
I think we started from one of the examples in HeterogeneousCore/AlpakaTest/plugins which uses the signature above.
Ok so I change this to return HGCalRecHitDevice using std::move(*ptr).

Reply:
The `device_recHits` does not need to be a `unique_ptr`, i.e. you could do
```cpp
HGCalRecHitDevice device_recHits(device_digis.view().metadata().size(), queue);
...
return device_recHits;
```

My _guess_ is that the present `std::move(*ptr)` results in move-construction of the returned `HGCalRecHitDevice`, whereas avoiding the `unique_ptr` should allow the copy elision @fwyzard was after.

Reply:
ok I totally misunderstood but at the same time I realized I was unaware of the intricacies of copy elision and that I need guidance as I don't fully understand what I find in the web
Would one of this options work?
- return `std::shared_ptr`
- return `std::unique_ptr` but use std::move(ptr) ?
- return `std::weak_ptr`
thanks a lot for the help

Reply:
None of them.

You should simply do
```c++
HGCalRecHitDevice device_recHits(device_digis.view().metadata().size(), queue);
...
return device_recHits;
```

Copy elision means that the call to
```
auto recHits = calibrator_.calibrate(queue, hostDigis, deviceCalibParamProvider, deviceConfigParamProvider);
```
will use for `recHits` the same object as `device_recHits`, without making any copies.

Reply:
ok thanks for the recipe. I understand the concept, what I was not fully capturing were all the cases for different c++. Also btw in the meanwhile I asked chatGPT which proposed (and wrongly) my option 2 .
Ok... changing the code as proposed

Reply:
> Also btw in the meanwhile I asked chatGPT which proposed (and wrongly) my option 2 .

Eh, the last answer I got about C++ was code that would immediately segmentation fault..